src/stream/src/executor/backfill/snapshot_backfill/vnode_stream.rs

```diff
@@ -0,0 +1,198 @@
+        let streams =
+            FuturesUnordered::from_iter(vnode_streams.into_iter().map(|(vnode, stream)| {
+                StreamWithVnode {
+                    stream: Box::pin(stream.peekable()),
+                    vnode,
+                }
+                .into_future()
+            }));
+        let ops = Vec::with_capacity(data_chunk_builder.batch_size());
+        Self {
+            streams,
+            finished_vnode: HashSet::new(),
+            data_chunk_builder,
+            ops,
+        }
+    }
+}
+
+impl<St: BackfillRowStream> VnodeStream<St> {
+    fn poll_next_row(
+        &mut self,
+        cx: &mut Context<'_>,
+    ) -> Poll<StreamExecutorResult<Option<BackfillRowItem>>> {
+        loop {
+            let ready_item = match ready!(self.streams.poll_next_unpin(cx)) {
+                None => Ok(None),
+                Some((None, stream)) => {
+                    assert!(self.finished_vnode.insert(stream.vnode));
+                    continue;
+                }
+                Some((Some(Ok(item)), stream)) => {
+                    self.streams.push(stream.into_future());
+                    Ok(Some(item))
+                }
+                Some((Some(Err(e)), _stream)) => Err(e),
+            };
+            break Poll::Ready(ready_item);
+        }
+    }
+
+    #[expect(dead_code)]
+    pub(super) fn consume_builder(&mut self) -> Option<StreamChunk> {
+        self.data_chunk_builder.consume_all().map(|chunk| {
+            let ops = replace(
+                &mut self.ops,
+                Vec::with_capacity(self.data_chunk_builder.batch_size()),
+            );
+            StreamChunk::from_parts(ops, chunk)
+        })
+    }
+
+    #[expect(dead_code)]
+    pub(super) async fn for_vnode_pk_progress(
+        &mut self,
+        pk_indices: &[usize],
+        mut on_vnode_progress: impl FnMut(VirtualNode, Option<OwnedRow>),
+    ) -> StreamExecutorResult<()> {
+        assert!(self.data_chunk_builder.is_empty());
+        for vnode in &self.finished_vnode {
+            on_vnode_progress(*vnode, None);
+        }
+        for vnode_stream in &mut self.streams {
+            let vnode_stream = vnode_stream.get_mut().expect("should exist");
+            match vnode_stream.stream.as_mut().peek().await {
+                Some(Ok(((_, row), extra))) => {
+                    let pk = row.project(pk_indices).to_owned_row();
+                    if cfg!(debug_assertions)
+                        && let Some((_, extra_row)) = extra
+                    {
+                        assert_eq!(pk, extra_row.project(pk_indices).to_owned_row());
+                    }
+                    on_vnode_progress(vnode_stream.vnode, Some(pk));
+                }
+                Some(Err(_)) => {
+                    return Err(vnode_stream
+                        .stream
+                        .try_next()
+                        .await
+                        .expect_err("checked Err"));
```

Comment:
Add a more descriptive error message?

Reply:
This is similar to `result.expect("...")`, but just expect that the result is `Err`.